lock/etcd.go

```diff
@@ -0,0 +1,88 @@
+package lock
+
+import (
+	"sync"
+
+	"github.com/coreos/etcd/clientv3"
+	"github.com/coreos/etcd/clientv3/concurrency"
+	"golang.org/x/net/context"
+)
+
+func NewEtcd(cfg clientv3.Config, ttl int) Locker {
+	l := &etcd{
+		cfg: cfg,
+		ttl: ttl,
+		m:   &sync.Mutex{},
+	}
+	go func() { _ = l.connect() }()
+	return l
+}
+
+type etcd struct {
+	cfg    clientv3.Config
+	ttl    int
+	client *clientv3.Client
+	m      *sync.Mutex
+}
+
+func (l *etcd) connect() error {
+	l.m.Lock()
+	defer l.m.Unlock()
+
+	if l.client == nil {
+		client, err := clientv3.New(l.cfg)
+		if err != nil {
+			return err
+		}
+
+		l.client = client
+	}
+
+	return nil
+}
+
+func (l *etcd) Lock(id string, ctx context.Context) (Lock, error) {
+	if err := l.connect(); err != nil {
+		return nil, err
+	}
+
+	session, err := concurrency.NewSession(l.client, concurrency.WithTTL(l.ttl))
+	if err != nil {
+		return nil, err
+	}
+
+	mutex := concurrency.NewMutex(session, id)
+	if err := mutex.Lock(ctx); err != nil {
+		_ = session.Close()
+		if cerr := ctx.Err(); cerr != nil {
+			return nil, ErrCanceled.Wrap(err)
+		}
+
+		return nil, err
+	}
+
+	return &etcdMutex{
+		m:       mutex,
+		session: session,
+		donec:   session.Done(),
+	}, nil
+}
+
+func (l *etcd) Close() error {
+	return l.client.Close()
+}
+
+type etcdMutex struct {
+	m       *concurrency.Mutex
+	session *concurrency.Session
+	donec   <-chan struct{}
+}
+
+func (m *etcdMutex) Done() <-chan struct{} {
+	return m.donec
+}
+
+func (m *etcdMutex) Unlock() {
+	_ = m.m.Unlock(context.TODO())
```

Comment:
Nope. Even if it returns an error, the lock will not be renewed and it will expire.